README.md

```diff
@@ -1,34 +1,741 @@
-# SumsUp
+# sums_up
 
-Sum types for Ruby.
-Inspired by [hojberg/sums-up](https://github.com/hojberg/sums-up).
+Sum types for Ruby with zero runtime dependencies. Inspired by [hojberg/sums-up](https://github.com/hojberg/sums-up).
 
-## Installation
+[![Build Status](https://travis-ci.org/nahiluhmot/sums_up.svg?branch=master)](https://travis-ci.org/nahiluhmot/sums_up)
 
-Add this line to your application's Gemfile:
+* [What is a Sum Type?](#what-is-a-sum-type)
+* [Quick Start](#quick-start)
+* [Defining Sum Types](#defining-sum-types)
+* [Predicates](#predicates)
+* [Pattern Matching with Hashes](#pattern-matching-with-hashes)
+* [Pattern Matching with Blocks](#pattern-matching-with-blocks)
+* [Methods on Sum Types](#methods-on-sum-types)
+* [Variant Instance Methods](#variant-instance-methods)
+* [A Note on Mutability](#a-note-on-mutability)
+* [Maybes](#maybes)
+* [Results](#results)
+* [Development](#development)
+* [Contributing](#contributing)
+* [License](#license)
+* [Code of Conduct](#code-of-conduct)
+
+## What is a Sum Type?
+
+Sum types are data structures with multiple variants.
+Ruby does not have sum types, but many concepts in the language (like booleans, integers, errors, etc.) can be described using sum types.
+Sum types are not limited to those use-cases, however, and are a powerful tool for modeling domain-specific data as well.
+
+This README will use non-generalized examples of sum types to help build an intuition for when they might be useful.
+To learn more about sum types, I recommend watching [Philip Wadler's Category Theory for the Working Hacker](https://www.youtube.com/watch?v=V10hzjgoklA) and checking out [Elm's Custom Types](https://guide.elm-lang.org/types/custom_types.html), [Haskell's Sum Types](https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/sum-types), and [Rust's Enums](https://doc.rust-lang.org/book/ch06-01-defining-an-enum.html).
+
```

Comment:
It's prolly worth a wikipedia link to adts as well.

Reply:
Also perhaps the fact that have a low max number of possible values by construction. This concept wasn't something I even thought about before using sum types.

Reply:
Could you expand a bit more on your second point?

Reply:
I mean the difference in "size" of a sum type vs a product type in how sum types are addition and product types are multiplication in their scope of how many possible values they can hold.

Reply:
Ah, yeah, that's a selling point for sure. I'll keep that in mind for future README updates.